Assets/Documentation/man/man3/FarbenButton.3
.TH "FarbenButton" 3 "Version 0.1" "NPP VR Simulation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
FarbenButton \- This class is used to change the color scheme in the scene to accommodate color blindness\&.  

.SH SYNOPSIS
.br
.PP
.PP
Inherits MonoBehaviour\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBChangeColorOfGazeGuidingButtons\fP ()"
.br
.RI "This method applies a new colour scheme to the \fBGazeGuidingButtons\fP in the scene\&. "
.ti -1c
.RI "void \fBrevertChangeColorOfGazeGuidingButtons\fP ()"
.br
.RI "This method reverts the colour scheme of the \fBGazeGuidingButtons\fP in the scene to the original colour scheme\&. "
.ti -1c
.RI "void \fBChangeMeltdownColorForLamps\fP ()"
.br
.RI "This method applies a new colour scheme to the lamps signalling the failure of specific components\&. "
.ti -1c
.RI "void \fBrevertChangeMeltdownColorForLamps\fP ()"
.br
.RI "This method reverts the colour scheme of the lamps signalling the failure of specific components to the original colour scheme\&. "
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "XRBaseInteractable \fBinteractable\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIinteractable\fP is a reference to a XRSimpleInteractable
.RE
.PP
"
.ti -1c
.RI "float \fBmoveDistance\fP = 0\&.5f"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fImoveDistance\fP specifies how far to move the button during the animation 
.RE
.PP
"
.ti -1c
.RI "float \fBanimationDuration\fP = 0\&.5f"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIanimationDuration\fP specifies the duration of the animation
.RE
.PP
"
.ti -1c
.RI "float \fBcooldownDuration\fP = 1f"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIcooldownDuration\fP specifies the duration of the animation cooldown
.RE
.PP
"
.ti -1c
.RI "Material[] \fBmaterialsEmissionChange\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fImaterialsEmissionChange\fP is an array of materials to adjust
.RE
.PP
"
.ti -1c
.RI "Color[] \fBnewColorsEmission\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fInewColorsEmission\fP is an array of new colors to apply to materials
.RE
.PP
"
.ti -1c
.RI "float \fBemissionIntensity\fP = 1\&.0f"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIemissionIntensity\fP stores the Intensity of the emission color
.RE
.PP
"
.ti -1c
.RI "TextMeshPro[] \fBtextMeshProObjects\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fItextMeshProObjects\fP is an array of TextMeshPro objects to adjust
.RE
.PP
"
.ti -1c
.RI "Color \fBnewTextColors\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fInewTextColors\fP is an array of new colors to apply to TextMeshPro objects
.RE
.PP
"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBStart\fP ()"
.br
.RI "This method finds, stores and deactivates all active display annotations\&. "
.ti -1c
.RI "void \fBOnEnable\fP ()"
.br
.RI "This method is called when the object is enabled and adds event listeners for the selectEntered and selectExited events\&. "
.ti -1c
.RI "void \fBOnDisable\fP ()"
.br
.RI "This method is called when the object is disabled and removes event listeners for the selectEntered and selectExited events\&. "
.ti -1c
.RI "void \fBOnSelectEntered\fP (SelectEnterEventArgs args)"
.br
.RI "This method is called when an interactor enters the object and toggles colour blind mode\&. "
.ti -1c
.RI "IEnumerator \fBAnimateTransform\fP ()"
.br
.RI "This method animates the button by moving it back and forth calling \fBMoveToPosition()\fP\&. "
.ti -1c
.RI "IEnumerator \fBMoveToPosition\fP (Vector3 start, Vector3 end, float duration)"
.br
.RI "This method is called in \fBAnimateTransform()\fP to move the button\&. "
.ti -1c
.RI "void \fBChangeMaterialsEmission\fP ()"
.br
.RI "This method applies a new emission colour scheme to all the materials in the scene\&. "
.ti -1c
.RI "void \fBrevertChangeMaterialsEmission\fP ()"
.br
.RI "This method reverts the emission colour scheme of all the materials in the scene to the original colour scheme\&. "
.ti -1c
.RI "void \fBChangeTextMeshProColors\fP ()"
.br
.RI "This method applies a new colour scheme to all TextMeshPro objects in the scene\&. "
.ti -1c
.RI "void \fBrevertChangeTextMeshProColors\fP ()"
.br
.RI "This method reverts the colour scheme of all TextMeshPro objects in the scene to the original colour scheme\&. "
.ti -1c
.RI "void \fBChangeCueColor\fP ()"
.br
.RI "This method applies a new colour scheme to the DirectionCue objects in the scene\&. "
.ti -1c
.RI "void \fBrevertChangeCueColor\fP ()"
.br
.RI "This method reverts the colour scheme of the DirectionCue objects in the scene back to the original color scheme\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "bool \fBisAnimating\fP = false"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIisAnimating\fP tracks whether the button is currently being animated
.RE
.PP
"
.ti -1c
.RI "Color[] \fBoriginalColors\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIoriginalColors\fP is an array storing original colours
.RE
.PP
"
.ti -1c
.RI "Color[] \fBoriginalEmissionColors\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIoriginalEmissionColors\fP is an array storing original emission colours
.RE
.PP
"
.ti -1c
.RI "Color[] \fBoriginalTextColors\fP"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIoriginalTextColors\fP is an array storing original text colours
.RE
.PP
"
.ti -1c
.RI "bool \fBcolorBlindModeActivated\fP = false"
.br
.RI "
.PP
\fBParameters\fP
.RS 4
\fIcolorBlindModeActivated\fP tracks whether the color blind mode is currently active
.RE
.PP
"
.in -1c
.SH "Detailed Description"
.PP 
This class is used to change the color scheme in the scene to accommodate color blindness\&. 
.SH "Member Function Documentation"
.PP 
.SS "IEnumerator FarbenButton\&.AnimateTransform ()\fR [private]\fP"

.PP
This method animates the button by moving it back and forth calling \fBMoveToPosition()\fP\&. 
.SS "void FarbenButton\&.ChangeColorOfGazeGuidingButtons ()"

.PP
This method applies a new colour scheme to the \fBGazeGuidingButtons\fP in the scene\&. 
.SS "void FarbenButton\&.ChangeCueColor ()\fR [private]\fP"

.PP
This method applies a new colour scheme to the DirectionCue objects in the scene\&. 
.SS "void FarbenButton\&.ChangeMaterialsEmission ()\fR [private]\fP"

.PP
This method applies a new emission colour scheme to all the materials in the scene\&. 
.SS "void FarbenButton\&.ChangeMeltdownColorForLamps ()"

.PP
This method applies a new colour scheme to the lamps signalling the failure of specific components\&. 
.SS "void FarbenButton\&.ChangeTextMeshProColors ()\fR [private]\fP"

.PP
This method applies a new colour scheme to all TextMeshPro objects in the scene\&. 
.SS "IEnumerator FarbenButton\&.MoveToPosition (Vector3 start, Vector3 end, float duration)\fR [private]\fP"

.PP
This method is called in \fBAnimateTransform()\fP to move the button\&. 
.PP
\fBParameters\fP
.RS 4
\fIstart\fP is a Vector3 specifying the origin of the movement
.br
\fIend\fP is a Vector3 specifying the end position of the movement
.br
\fIduration\fP specifies the duration of the animation
.RE
.PP

.SS "void FarbenButton\&.OnDisable ()\fR [private]\fP"

.PP
This method is called when the object is disabled and removes event listeners for the selectEntered and selectExited events\&. Additionally the original colour scheme is restored\&. 
.SS "void FarbenButton\&.OnEnable ()\fR [private]\fP"

.PP
This method is called when the object is enabled and adds event listeners for the selectEntered and selectExited events\&. 
.SS "void FarbenButton\&.OnSelectEntered (SelectEnterEventArgs args)\fR [private]\fP"

.PP
This method is called when an interactor enters the object and toggles colour blind mode\&. 
.PP
\fBParameters\fP
.RS 4
\fIargs\fP passes event specific arguments upon entering the interaction
.RE
.PP

.SS "void FarbenButton\&.revertChangeColorOfGazeGuidingButtons ()"

.PP
This method reverts the colour scheme of the \fBGazeGuidingButtons\fP in the scene to the original colour scheme\&. 
.SS "void FarbenButton\&.revertChangeCueColor ()\fR [private]\fP"

.PP
This method reverts the colour scheme of the DirectionCue objects in the scene back to the original color scheme\&. 
.SS "void FarbenButton\&.revertChangeMaterialsEmission ()\fR [private]\fP"

.PP
This method reverts the emission colour scheme of all the materials in the scene to the original colour scheme\&. 
.SS "void FarbenButton\&.revertChangeMeltdownColorForLamps ()"

.PP
This method reverts the colour scheme of the lamps signalling the failure of specific components to the original colour scheme\&. 
.SS "void FarbenButton\&.revertChangeTextMeshProColors ()\fR [private]\fP"

.PP
This method reverts the colour scheme of all TextMeshPro objects in the scene to the original colour scheme\&. 
.SS "void FarbenButton\&.Start ()\fR [private]\fP"

.PP
This method finds, stores and deactivates all active display annotations\&. It also stores the original colour scheme of the scene\&. 
.SH "Member Data Documentation"
.PP 
.SS "float FarbenButton\&.animationDuration = 0\&.5f"

.PP

.PP
\fBParameters\fP
.RS 4
\fIanimationDuration\fP specifies the duration of the animation
.RE
.PP

.SS "bool FarbenButton\&.colorBlindModeActivated = false\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIcolorBlindModeActivated\fP tracks whether the color blind mode is currently active
.RE
.PP

.SS "float FarbenButton\&.cooldownDuration = 1f"

.PP

.PP
\fBParameters\fP
.RS 4
\fIcooldownDuration\fP specifies the duration of the animation cooldown
.RE
.PP

.SS "float FarbenButton\&.emissionIntensity = 1\&.0f"

.PP

.PP
\fBParameters\fP
.RS 4
\fIemissionIntensity\fP stores the Intensity of the emission color
.RE
.PP

.SS "XRBaseInteractable FarbenButton\&.interactable"

.PP

.PP
\fBParameters\fP
.RS 4
\fIinteractable\fP is a reference to a XRSimpleInteractable
.RE
.PP

.SS "bool FarbenButton\&.isAnimating = false\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIisAnimating\fP tracks whether the button is currently being animated
.RE
.PP

.SS "Material [] FarbenButton\&.materialsEmissionChange"

.PP

.PP
\fBParameters\fP
.RS 4
\fImaterialsEmissionChange\fP is an array of materials to adjust
.RE
.PP

.SS "float FarbenButton\&.moveDistance = 0\&.5f"

.PP

.PP
\fBParameters\fP
.RS 4
\fImoveDistance\fP specifies how far to move the button during the animation 
.RE
.PP

.SS "Color [] FarbenButton\&.newColorsEmission"

.PP

.PP
\fBParameters\fP
.RS 4
\fInewColorsEmission\fP is an array of new colors to apply to materials
.RE
.PP

.SS "Color FarbenButton\&.newTextColors"

.PP

.PP
\fBParameters\fP
.RS 4
\fInewTextColors\fP is an array of new colors to apply to TextMeshPro objects
.RE
.PP

.SS "Color [] FarbenButton\&.originalColors\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIoriginalColors\fP is an array storing original colours
.RE
.PP

.SS "Color [] FarbenButton\&.originalEmissionColors\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIoriginalEmissionColors\fP is an array storing original emission colours
.RE
.PP

.SS "Color [] FarbenButton\&.originalTextColors\fR [private]\fP"

.PP

.PP
\fBParameters\fP
.RS 4
\fIoriginalTextColors\fP is an array storing original text colours
.RE
.PP

.SS "TextMeshPro [] FarbenButton\&.textMeshProObjects"

.PP

.PP
\fBParameters\fP
.RS 4
\fItextMeshProObjects\fP is an array of TextMeshPro objects to adjust
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for NPP VR Simulation from the source code\&.


Assets/Skripte/UI/FarbenButton.cs
using UnityEngine;
using System.Collections;
using System.Linq;
using UnityEngine.XR.Interaction.Toolkit;
using UnityEngine.XR.Interaction.Toolkit.Interactables;
using TMPro;
using UnityEditor;
using UnityEngine.UI;

/// <summary>
/// This class is used to change the color scheme in the scene to accommodate color blindness.
/// </summary>
public class FarbenButton : MonoBehaviour
{
    /// <param name="interactable"> is a reference to a XRSimpleInteractable</param>
    public XRBaseInteractable interactable;
    /// <param name="moveDistance"> specifies how far to move the button during the animation </param>
    public float moveDistance = 0.5f;            
    /// <param name="animationDuration"> specifies the duration of the animation</param>
    public float animationDuration = 0.5f;
    /// <param name="cooldownDuration"> specifies the duration of the animation cooldown</param>
    public float cooldownDuration = 1f;
    /// <param name="materialsEmissionChange"> is an array of materials to adjust</param>
    public Material[] materialsEmissionChange;  
    /// <param name="newColorsEmission"> is an array of new colors to apply to materials</param>
    public Color[] newColorsEmission;           
    /// <param name="emissionIntensity"> stores the Intensity of the emission color</param>
    public float emissionIntensity = 1.0f;      
    /// <param name="textMeshProObjects"> is an array of TextMeshPro objects to adjust</param>
    public TextMeshPro[] textMeshProObjects;    
    /// <param name="newTextColors"> is an array of new colors to apply to TextMeshPro objects</param>
    public Color newTextColors;                 
    /// <param name="isAnimating"> tracks whether the button is currently being animated</param>
    private bool isAnimating = false;
    /// <param name="originalColors"> is an array storing original colours</param>
    private Color[] originalColors;
    /// <param name="originalEmissionColors"> is an array storing original emission colours</param>
    private Color[] originalEmissionColors;
    /// <param name="originalTextColors"> is an array storing original text colours</param>
    private Color[] originalTextColors;

    /// <summary>
    /// This method finds, stores and deactivates all active display annotations. It also stores the original colour scheme of the scene.
    /// </summary>
    private void Start()
    {
        // Find all GameObjects named "AnzeigenMarker"
        // Find all GameObjects named "AnzeigenMarker"
        GameObject[] anzeigenMarkers = Resources.FindObjectsOfTypeAll<GameObject>()
            .Where(go => go.name == "AnzeigenMarker" 
        #if UNITY_EDITOR
                         && !AssetDatabase.Contains(go)
        #endif
            )
            .ToArray();

        // Initialize textMeshProObjects with TextMeshPro components of children named "AusrufeZeichen"
        textMeshProObjects = Resources.FindObjectsOfTypeAll<TextMeshPro>()
            .Where(tmp => tmp.gameObject.name == "AusrufeZeichen")
            .ToArray();

        // Deactivate all "AnzeigenMarker" objects
        foreach (var marker in anzeigenMarkers)
        {
            marker.SetActive(false);
        }
        
        // Save original colors
        originalColors = new Color[materialsEmissionChange.Length];
        originalEmissionColors = new Color[materialsEmissionChange.Length];
        for (int i = 0; i < materialsEmissionChange.Length; i++)
        {
            if (materialsEmissionChange[i] != null)
            {
                originalColors[i] = materialsEmissionChange[i].color;
                if (materialsEmissionChange[i].IsKeywordEnabled("_EMISSION"))
                {
                    originalEmissionColors[i] = materialsEmissionChange[i].GetColor("_EmissionColor");
                }
            }
        }

        // Save original text colors
        originalTextColors = new Color[textMeshProObjects.Length];
        for (int i = 0; i < textMeshProObjects.Length; i++)
        {
            if (textMeshProObjects[i] != null)
            {
                originalTextColors[i] = textMeshProObjects[i].color;
            }
        }
    }
    

    /// <summary>
    /// This method is called when the object is enabled and adds event listeners for the selectEntered and selectExited events.
    /// </summary>    
    private void OnEnable()
    {
        var interactable = GetComponent<XRSimpleInteractable>();
        interactable.selectEntered.AddListener(OnSelectEntered);
    }

    /// <summary>
    /// This method is called when the object is disabled and removes event listeners for the selectEntered and selectExited events. Additionally the original colour scheme is restored.
    /// </summary>
    private void OnDisable()
    {
        var interactable = GetComponent<XRSimpleInteractable>();
        interactable.selectEntered.RemoveListener(OnSelectEntered);

        // Restore original colors
        for (int i = 0; i < materialsEmissionChange.Length; i++)
        {
            if (materialsEmissionChange[i] != null && i < originalColors.Length && i < originalEmissionColors.Length)
            {
                materialsEmissionChange[i].color = originalColors[i];
                if (materialsEmissionChange[i].IsKeywordEnabled("_EMISSION"))
                {
                    materialsEmissionChange[i].SetColor("_EmissionColor", originalEmissionColors[i]);
                }
            }
        }

        // Restore original text colors
        for (int i = 0; i < textMeshProObjects.Length; i++)
        {
            if (textMeshProObjects[i] != null && i < originalTextColors.Length)
            {
                textMeshProObjects[i].color = originalTextColors[i];
            }
        }
    }

    /// <param name="colorBlindModeActivated"> tracks whether the color blind mode is currently active</param>
    private bool colorBlindModeActivated = false;

    /// <summary>
    /// This method is called when an interactor enters the object and toggles colour blind mode.
    /// </summary>
    /// <param name="args"> passes event specific arguments upon entering the interaction</param>    
    private void OnSelectEntered(SelectEnterEventArgs args)
    {
        if (!isAnimating && !colorBlindModeActivated)
        {
            StartCoroutine(AnimateTransform());
            ChangeMaterialsEmission();
            ChangeTextMeshProColors();
            ChangeCueColor();
            ChangeColorOfGazeGuidingButtons();
            ChangeMeltdownColorForLamps();
            colorBlindModeActivated = true;
        }else if (!isAnimating && colorBlindModeActivated)
        {
            // Restore original colors
            
            StartCoroutine(AnimateTransform());
            revertChangeMaterialsEmission();
            revertChangeTextMeshProColors();
            revertChangeCueColor();
            revertChangeColorOfGazeGuidingButtons();
            revertChangeMeltdownColorForLamps();
            colorBlindModeActivated = false;
        }
    }

    /// <summary>
    /// This method animates the button by moving it back and forth calling MoveToPosition().
    /// </summary>
    private IEnumerator AnimateTransform()
    {
        isAnimating = true;
        Vector3 initialPosition = transform.localPosition;
        Vector3 targetPosition = initialPosition + new Vector3(moveDistance, 0, 0);

        // Move to target position
        yield return MoveToPosition(initialPosition, targetPosition, animationDuration);

        // Move back to initial position
        yield return MoveToPosition(targetPosition, initialPosition, animationDuration);

        // Cooldown
        yield return new WaitForSeconds(cooldownDuration);

        isAnimating = false;
    }

    /// <summary>
    /// This method is called in AnimateTransform() to move the button.
    /// </summary>
    /// <param name="start"> is a Vector3 specifying the origin of the movement</param>
    /// <param name="end"> is a Vector3 specifying the end position of the movement</param>
    /// <param name="duration"> specifies the duration of the animation</param>
    private IEnumerator MoveToPosition(Vector3 start, Vector3 end, float duration)
    {
        float elapsedTime = 0f;
        while (elapsedTime < duration)
        {
            transform.localPosition = Vector3.Lerp(start, end, elapsedTime / duration);
            elapsedTime += Time.deltaTime;
            yield return null;
        }
        transform.localPosition = end;
    }

    /// <summary>
    /// This method applies a new emission colour scheme to all the materials in the scene.
    /// </summary>
    private void ChangeMaterialsEmission()
    {
        for (int i = 0; i < materialsEmissionChange.Length; i++)
        {
            if (materialsEmissionChange[i] != null && i < newColorsEmission.Length)
            {
                materialsEmissionChange[i].color = newColorsEmission[i];
                if (materialsEmissionChange[i].IsKeywordEnabled("_EMISSION"))
                {
                    materialsEmissionChange[i].SetColor("_EmissionColor", newColorsEmission[i] * emissionIntensity);
                }
            }
        }
    }

    /// <summary>
    /// This method reverts the emission colour scheme of all the materials in the scene to the original colour scheme.
    /// </summary>    
    private void revertChangeMaterialsEmission()
    {
        for (int i = 0; i < materialsEmissionChange.Length; i++)
        {
            if (materialsEmissionChange[i] != null && i < originalColors.Length && i < originalEmissionColors.Length)
            {
                materialsEmissionChange[i].color = originalColors[i];
                if (materialsEmissionChange[i].IsKeywordEnabled("_EMISSION"))
                {
                    materialsEmissionChange[i].SetColor("_EmissionColor", originalEmissionColors[i]);
                }
            }
        }
    }

    /// <summary>
    /// This method applies a new colour scheme to all TextMeshPro objects in the scene.
    /// </summary>
    private void ChangeTextMeshProColors()
    {
        for (int i = 0; i < textMeshProObjects.Length; i++)
        {
            if (textMeshProObjects[i] != null)
            {
                textMeshProObjects[i].color = newTextColors;
            }
        }
    }

    /// <summary>
    /// This method reverts the colour scheme of all TextMeshPro objects in the scene to the original colour scheme.
    /// </summary>    
    private void revertChangeTextMeshProColors()
    {
        for (int i = 0; i < textMeshProObjects.Length; i++)
        {
            if (textMeshProObjects[i] != null)
            {
                textMeshProObjects[i].color = Color.red;
            }
        }
    }
    
    /// <summary>
    /// This method applies a new colour scheme to the DirectionCue objects in the scene.
    /// </summary>
    private void ChangeCueColor()
    {
        // Load the new sprite from the Resources folder
        Sprite newSprite = Resources.Load<Sprite>("DirectionCueMagenta");

        if (newSprite == null)
        {
            Debug.LogError("DirectionCueMagenta.png not found in Resources folder");
            return;
        }

        // Find all objects with Image components named DirectionCue and DirectionCue2
        Image[] directionCues = FindObjectsByType<Image>(FindObjectsSortMode.None)
            .Where(img => img.gameObject.name == "DirectionCue" || img.gameObject.name == "DirectionCue2")
            .ToArray();

        // Swap the image with the new sprite
        foreach (var img in directionCues)
        {
            img.sprite = newSprite;
        }
    }

    /// <summary>
    /// This method reverts the colour scheme of the DirectionCue objects in the scene back to the original color scheme.
    /// <summary>
    private void revertChangeCueColor()
    {
        Sprite newSprite = Resources.Load<Sprite>("DirectionCue");

        if (newSprite == null)
        {
            Debug.LogError("DirectionCue.png not found in Resources folder");
            return;
        }

        // Find all objects with Image components named DirectionCue and DirectionCue2
        Image[] directionCues = FindObjectsByType<Image>(FindObjectsSortMode.None)
            .Where(img => img.gameObject.name == "DirectionCue" || img.gameObject.name == "DirectionCue2")
            .ToArray();

        // Swap the image with the new sprite
        foreach (var img in directionCues)
        {
            img.sprite = newSprite;
        }
    }

    /// <summary>
    /// This method applies a new colour scheme to the GazeGuidingButtons in the scene.
    /// </summary>
    public void ChangeColorOfGazeGuidingButtons()
    {
        Flipper[] flippers = FindObjectsByType<Flipper>(FindObjectsSortMode.None);
        foreach (var flipper in flippers)
        {
            flipper.updateMaterials();
        }
    }

    /// <summary>
    /// This method reverts the colour scheme of the GazeGuidingButtons in the scene to the original colour scheme.
    /// </summary>    
    public void revertChangeColorOfGazeGuidingButtons()
    {
        Flipper[] flippers = FindObjectsByType<Flipper>(FindObjectsSortMode.None);
        foreach (var flipper in flippers)
        {
            flipper.revertMaterials();
        }
    }

    /// <summary>
    /// This method applies a new colour scheme to the lamps signalling the failure of specific components.
    /// </summary>
    public void ChangeMeltdownColorForLamps()
    {
        AusfallAnzeigenManager ausfallAnzeigenManager = FindAnyObjectByType<AusfallAnzeigenManager>();
        ausfallAnzeigenManager.toChange = Color.magenta;
    }

    /// <summary>
    /// This method reverts the colour scheme of the lamps signalling the failure of specific components to the original colour scheme.
    /// </summary>    
    public void revertChangeMeltdownColorForLamps()
    {
        AusfallAnzeigenManager ausfallAnzeigenManager = FindAnyObjectByType<AusfallAnzeigenManager>();
        ausfallAnzeigenManager.toChange = Color.white;
    }
}